### ISSUE TITLE
issue opening multiple files at once within a Savefile archive.

### ISSUE THREAD

**Issue by CaitSith2** [Thealexbarney/libhac#4]
It seems to be unsafe to open more than one file within a savefile at a time.  The moment I open more than one, the stream position of the first file gets messed up badly enough to cause exceptions to be thrown.  For example, trying to read tickets more efficiently than by brute force reading.

```csharp
using (var stream = File.Open(Path.Combine(Configuration.Data.SystemPath, "save", "80000000000000e1"), FileMode.Open, FileAccess.Read))
{
    var commonTickets = new Savefile(stream);
    var ticketList = new BinaryReader(commonTickets.OpenFile("/ticket_list.bin"));
    var tickets = new BinaryReader(commonTickets.OpenFile("/ticket.bin"));  //Opening this file
    ulong titleID;
    do
    {
        titleID = ticketList.ReadUInt64();  //causes this read to fail with an exception thrown.
        if (titleID == ulong.MaxValue) continue;
        ticketList.BaseStream.Position += 0x18;
        _tickets[$"{titleID:x16}"] = new Ticket(tickets.ReadBytes(0x400));
    } while (titleID != ulong.MaxValue);
}
```

**Comment by Thealexbarney** [comment]
I've added some stuff in the past few days to allow sharing a stream and avoid that situation. It's been applied to nca and romfs files, but I forgot that it would apply to save files too. Guess I'll do that today

**Comment by CaitSith2** [comment]
Also take note that you can just as easily take this code to more efficiently extract the tickets. 

ticket_list.bin is the following format.

Length | Content
--- | ---
16 bytes | Rights ID
8 bytes | Ticket ID
4 bytes | Account ID
4 bytes | reserved

Repeat reading until you encounter 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF00000000.  This tells you how many tickets you have.  From there, each 1KiB in ticket.bin is the matching ticket.

As it is,  file 80000000000000e3 is interesting in that it doesn't contains tickets, but rather proof of pre-order purchase info.  In this case, the information is stored in ticket_list.bin, in the following format.

Length | Content
--- | ---
16 bytes | Rights ID (master key is set to 0, even if it is something else normally)
8 bytes | Ticket ID
4 bytes | Account ID
4 bytes | reserved
8 bytes | Release date/time in UTC seconds since Jan 1, 1970.
216 bytes | reserved.

Likewise, this data repeats until a 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF00000000 is encountered.

**Comment by Thealexbarney** [comment]
Fixed in 80cf0b4, and I've updated the title key dumpers to use the new savefile reading capability instead of brute-force searching in 1dc232e.

**Comment by Thealexbarney** [comment]
@CaitSith2 On the topic of ticket reading, if you target .NET Framework 4.6+ you can use the built-in RSA-OAEP to decrypt the title key. It's much faster than doing the calculation with BigInteger.
[Example](https://github.com/Thealexbarney/libhac/blob/d7929fc45863dc8860355f3dc4fe7bf581759709/libhac/Crypto.cs#L78-L284)

You'll have to recover some of the RSA parameters from the modulus and exponents before .NET will use the key.

**Comment by Thealexbarney** [comment]
Something seems off with those numbers. The difference between the two sounds right, but on my computer it took ~50ms to decrypt each title key with BigInteger. With the .NET crypto APIs it takes <1ms per key.

Yep, it's RSA-OAEP using SHA256. I only yesterday found out that they added that capability to .NET 4.6

**Comment by CaitSith2** [comment]
I have confirmed that some of my timings is consistent with other operations that happen at the same time, but even when minimized to the absolute minimum of read ticket, decrypt title key,  the speed of the operation on my system is ~160ms per key old method, and ~1.5 per key using the native method,

One thing I can't seem to get is micro sd cards to have unique keys on the same console.  I wonder if nintendo silently changed that somewhere along the line, so that each SD card that goes in the console actually ends up with the SAME SD encryption seed.  (The private file on ALL of the SD cards I have inserted that have NEVER been inserted previously into that switch ALL end up with the same 16 bytes.), which would result in the exact same 16 byte SD seed, when looked for in save file 80....43.  (It does seem the seed will change on a console format, although that doesn't happen often.)

Also, wouldn't mind finding some other way to communicate beside reusing an old issue for this purpose.
